### PR TITLE
Upgrade OSSEC to v3.6.0

### PR DESCRIPTION
**Pull request by zenmonkeykstop**
## Status

Ready for review

## Description of Changes

Upgrades OSSEC to version 3.6.0.

### staging environment

(assuming the use of libvirt VMs)
- checkout this branch and run:

```
make build-debs
make staging
```

- [ ] `make staging` completes successfully
- log into the app server with `molecule login -s libvirt-staging-xenial -h app-staging`
  - [ ] `sudo systemctl status ossec.service` shows that the ossec agent and daemons are running
  - [ ] the log at `/var/ossec/logs/ossec.log` shows no unexpected errors
  - [ ] there are no grsec denies logged in `/var/log/kern.log` related to ossec binaries
- log into the monitor server with `molecule login -s libvirt-staging-xenial -h mon-staging`
  - [ ] `sudo systemctl status ossec.service` shows that the ossec agent and daemons are running
  - [ ] the log at `/var/ossec/logs/ossec.log` shows no unexpected errors
  - [ ] the log at `/var/ossec/logs/alerts/alerts.log` shows alerts being generated for both `mon-staging` and `app-staging`.
  - [ ] there are no grsec denies logged in `/var/log/kern.log` related to ossec binaries

### production (HW or VMs)

- checkout this branch and run `make build-debs`.
- copy the ossec VMs in `build/xenial` to the monitor server (`securedrop-ossec-server*.deb` and `ossec-server*.deb`) and application server  (`securedrop-ossec-agent*.deb` and `ossec-agent*.deb`) and install them.
- [ ]  The OSSEC server and agent services start successfully on the monitor and application server respectively.
- [ ] OSSEC emails start to flow as expected, encrypted with the key specified during installation.
- [ ] The test OSSEC email feature in the Journalist Interface works as expected.

## Deployment
- Will be deployed to new instances as part of installation, and old instances with next upgrade. 

### If you made changes to the system configuration:

- [x] [Configuration tests](https://docs.securedrop.org/en/latest/development/testing_configuration_tests.html) pass

### If you made non-trivial code changes:

- [x] I have written a test plan and validated it for this PR

### CODE REVIEW

**Comment by emkll** [comment]
I understand this is not ready for final review, but took a pass through it (clean install and upgrade install using prod VMs (to ensure email works).

The upgrade on prod VMs worked as expected (though more testing will be performed as part of the final review): emails were flowing and log files didn't show any errors.

A couple of things i've observed during the staging clean install:
### re-registration
The registration seems to no longer be idempotent, based on my local testing. Can you reproduce the following error output by running`make build-debs, make staging,make staging`?
```
    TASK [ossec : Register OSSEC agent.] *******************************************
    fatal: [app-staging]: FAILED! => {"changed": true, "cmd": ["/var/ossec/bin/agent-auth", "-m", "10.0.1.3", "-p", "1515", "-A", "app-staging", "-P", "/var/ossec/etc/authd.pass"], "delta": "0:00:01.020907", "end": "2020-04-17 16:30:30.387493", "failed_when_result": true, "rc": 0, "start": "2020-04-17 16:30:29.366586", "stderr": "2020/04/17 16:30:29 ossec-authd: INFO: Started (pid: 10782).\n2020/04/17 16:30:29 INFO: Connected to 10.0.1.3 at address 10.0.1.3, port 1515", "stderr_lines": ["2020/04/17 16:30:29 ossec-authd: INFO: Started (pid: 10782).", "2020/04/17 16:30:29 INFO: Connected to 10.0.1.3 at address 10.0.1.3, port 1515"], "stdout": "INFO: Using specified password.\nINFO: Connected to 10.0.1.3:1515\nINFO: Using agent name as: app-staging\nINFO: Send request to manager. Waiting for reply.\nERROR: Invalid IP address: 1024 (from manager)\nERROR: Unable to add agent. (from manager)\nERROR: Unable to create key. Either wrong password or connection not accepted by the manager.\nINFO: Connection closed.", "stdout_lines": ["INFO: Using specified password.", "INFO: Connected to 10.0.1.3:1515", "INFO: Using agent name as: app-staging", "INFO: Send request to manager. Waiting for reply.", "ERROR: Invalid IP address: 1024 (from manager)", "ERROR: Unable to add agent. (from manager)", "ERROR: Unable to create key. Either wrong password or connection not accepted by the manager.", "INFO: Connection closed."]}
```

Upon closer inspection on the `mon` server, the error message returned might be misleading, 1024 was likely the agent id for app-staging, these are the relevant logs in ossec.log on mon-staging:
```
2020/04/17 16:30:28 ossec-authd: INFO: Started (pid: 7196).
2020/04/17 16:30:28 Accepting connections. Using password specified on file: /var/ossec/etc/authd.pass
2020/04/17 16:30:28 IPv4: 0.0.0.0 on port 1515
2020/04/17 16:30:28 Request for TCP listen() succeeded.
2020/04/17 16:30:28 Socket bound for IPv4: 0.0.0.0 on port 1515
2020/04/17 16:30:29 ossec-authd: INFO: New connection from 10.0.1.2
2020/04/17 16:30:29 ossec-authd: INFO: Received request for a new agent (app-staging) from: 10.0.1.2
2020/04/17 16:30:29 ossec-authd: ERROR: Invalid IP address 1024 (duplicated)
```

### pax flags and decoder errors
This one is a bit confusing to me, because it only happens on clean install and was not observed on the upgrade install

Based on my local testing, it's unclear if the paxflags are required on app and mon. I removed the functions in the postinst that applied the paxflags (however as a result, mon did have paxctld installed with default configuration). I did not observe any grsec errors in kern.log or syslog, but did observe some ossec errors on mon while starting the service:

```
2020/04/17 16:10:18 Accepting connections. Using password specified on file: /var/ossec/etc/authd.pass
2020/04/17 16:10:18 IPv4: 0.0.0.0 on port 1515
2020/04/17 16:10:18 Request for TCP listen() succeeded.
2020/04/17 16:10:18 Socket bound for IPv4: 0.0.0.0 on port 1515
2020/04/17 16:10:19 ossec-authd: INFO: New connection from 10.0.1.2
2020/04/17 16:10:19 ossec-authd: INFO: Received request for a new agent (app-staging) from: 10.0.1.2
2020/04/17 16:10:19 ossec-authd: INFO: Agent key generated for app-staging (requested by 10.0.1.2)
2020/04/17 16:10:19 ossec-authd: INFO: Agent key created for app-staging (requested by 10.0.1.2)
2020/04/17 16:10:20 ossec-authd(1225): INFO: SIGNAL [(15)-(Terminated)] Received. Exit Cleaning...
2020/04/17 16:10:20 ossec-analysisd(1450): ERROR: Syntax error on regex: '\(pam_unix\)$': 9.
2020/04/17 16:10:20 ossec-testrule(1202): ERROR: Configuration error at '/etc/decoder.xml'. Exiting.
2020/04/17 16:12:20 ossec-analysisd(1450): ERROR: Syntax error on regex: '\(pam_unix\)$': 9.
2020/04/17 16:12:20 ossec-testrule(1202): ERROR: Configuration error at '/etc/decoder.xml'. Exiting.
2020/04/17 16:16:43 ossec-analysisd(1450): ERROR: Syntax error on regex: '\(pam_unix\)$': 9.
2020/04/17 16:16:43 ossec-testrule(1202): ERROR: Configuration error at '/etc/decoder.xml'. Exiting.
2020/04/17 16:16:43 ossec-analysisd(1450): ERROR: Syntax error on regex: '\(pam_unix\)$': 9.
2020/04/17 16:16:43 ossec-testrule(1202): ERROR: Configuration error at '/etc/decoder.xml'. Exiting.
2020/04/17 16:16:43 ossec-analysisd(1450): ERROR: Syntax error on regex: '\(pam_unix\)$': 9.
2020/04/17 16:16:43 ossec-testrule(1202): ERROR: Configuration error at '/etc/decoder.xml'. Exiting.
2020/04/17 16:16:44 ossec-analysisd(1450): ERROR: Syntax error on regex: '\(pam_unix\)$': 9.
2020/04/17 16:16:44 ossec-testrule(1202): ERROR: Configuration error at '/etc/decoder.xml'. Exiting.
2020/04/17 16:16:44 ossec-analysisd(1450): ERROR: Syntax error on regex: '\(pam_unix\)$': 9.
2020/04/17 16:16:44 ossec-testrule(1202): ERROR: Configuration error at '/etc/decoder.xml'. Exiting.
2020/04/17 16:16:44 ossec-analysisd(1450): ERROR: Syntax error on regex: '\(pam_unix\)$': 9.
2020/04/17 16:16:44 ossec-testrule(1202): ERROR: Configuration error at '/etc/decoder.xml'. Exiting.
2020/04/17 16:16:44 ossec-analysisd(1450): ERROR: Syntax error on regex: '\(pam_unix\)$': 9.
2020/04/17 16:16:44 ossec-testrule(1202): ERROR: Configuration error at '/etc/decoder.xml'. Exiting.
2020/04/17 16:16:44 ossec-analysisd(1450): ERROR: Syntax error on regex: '\(pam_unix\)$': 9.
2020/04/17 16:16:44 ossec-testrule(1202): ERROR: Configuration error at '/etc/decoder.xml'. Exiting.
2020/04/17 16:20:40 ossec-analysisd(1450): ERROR: Syntax error on regex: '\(pam_unix\)$': 9.
2020/04/17 16:20:40 ossec-testrule(1202): ERROR: Configuration error at '/etc/decoder.xml'. Exiting.
```

**Comment by zenmonkeykstop** [comment]
Yup, it's confusing to me that without pax flags set (via either method), the binaries error out when they try to parse their first regex, but that there *isn't* a grsec error thrown in the logs. Will try to repro the registration issue above first, but after that it may be worth trying doing the ossec build with a local source build libpcre2 with JIT disabled, to try narrow things down.

**Comment by zenmonkeykstop** [comment]
I didn't replicate the re-registration error above. On second run of `make staging`, I see
```    
    TASK [ossec : Check whether Application Server is registered as OSSEC agent.] ***
    ok: [mon-staging]
```
and then the other tasks don't run.

**Comment by zenmonkeykstop** [comment]
Pushed an update to this with PCRE2 JIT disabled in the ossec build - this will apparently be slower but also removes the need for pax flags to be set.

File: install_files/securedrop-ossec-server/DEBIAN/postinst [hunk 1]
```diff
@@ -18,6 +18,7 @@ set -x
 
 case "$1" in
     configure)
+
```
[the remark this answers is not in the document]
Comment (zenmonkeykstop) [reply install_files/securedrop-ossec-server/DEBIAN/postinst:21]: nope, fixed.

File: install_files/securedrop-ossec-server/usr/share/doc/securedrop-ossec-server/changelog.Debian [hunk 2]
```diff
@@ -1,3 +1,9 @@
+securedrop-ossec-agent (3.6.0) unstable; urgency=low
+
+  * Upgrade to ossec 3.60
```
[the remark this answers is not in the document]
Comment (zenmonkeykstop) [reply install_files/securedrop-ossec-server/usr/share/doc/securedrop-ossec-server/changelog.Debian:3]: :heavy_check_mark:

File: install_files/ansible-base/roles/build-ossec-deb-pkg/defaults/main.yml [hunk 3]
```diff
@@ -22,4 +22,4 @@ ossec_build_rsync_generic_opts:
 ossec_build_rsync_ansible_hack_opt:
   - "--rsync-path='sudo rsync'"
 
-ossec_source_checksum: sha256:a271d665ed502b3df4ff055a177159dfc0bc8a69dd44eab1f7c57fe8fff42a98
+ossec_source_checksum: sha256:653828a19137b8a7e98af65e873318f7bb48137fe1e61b80577e13c316e04708
```
[the remark this answers is not in the document]
Comment (zenmonkeykstop) [reply install_files/ansible-base/roles/build-ossec-deb-pkg/defaults/main.yml:25]: yup, check's still there, GPG key is unchanged.

**Comment by emkll** [comment]
I clicked the merge button a couple of hours ago , but due to GitHub issues/outage, it seems like this PR was merged into develop in https://github.com/freedomofpress/securedrop/commit/ab8d4a09f6b6a633398910ae4e008a5a316885c2 . However this PR is not displaying as `merged`

**Comment by zenmonkeykstop** [comment]
Confirmed that the merge commit and pr commits are present on `develop` - gonna close this PR to avoid confusion.

**Comment by zenmonkeykstop** [comment]
Lol and now it shows as merged :roll_eyes:

**Pull request closed without merging**